### PR TITLE
[GTK] Correct scaling for DPI.

### PR DESCRIPTION
**Pull request by gwhitney**
#### 69ad530f10603697a1c53b063840abc8ba7d9efe
<pre>
[GTK] Correct scaling for DPI.
<a href="https://bugs.webkit.org/show_bug.cgi?id=247980">https://bugs.webkit.org/show_bug.cgi?id=247980</a>

Reviewed by NOBODY (OOPS!).

Move the scaling factor formerly in WebKitWebView.cpp (responsive to
the xft-dpi setting) into WebKitWebViewBase.cpp, and split it into two
scaling factors. The first, a page scaling factor, uses screenDPI to
ensure that a length specified by a CSS unit of 1in will appear on the
screen as one physical inch. The second, a text scaling factor, ensures
that text specified as 96px in size will have a height on-screen as
specified by fontDPI (which is primarily driven by the xft-dpi setting),
in terms of the number of screen pixels taken up.

The two scaling factors are refreshed when any one of the inputs for
computing them changes: the GDK_SCALE factor, the xft-dpi setting, or
the screenDPI of the device it is displayed on (say by being dragged
to a different monitor). Throughout, at the default zoom level, a CSS
1in dimension is kept at one physical inch.

The scale factors are kept internal to WebKitWebViewBase (which helps
keep GTK dependencies grouped and reduces the total code in WebKitWebView
enclosed in `#ifdef PLATFORM(GTK)`) but made available to WebKitWebView
so that its external interfaces can maintain zoom level 1 as the default
zoom level, and zoom in and out from there.

So indeed, if a user of WebKitWebView asks for a zoom level of 1.5, a
length specified in CSS as 1in will then occupy one and a half physical
inches on screen.

* Source/WebCore/platform/gtk/PlatformScreenGtk.cpp:
(WebCore::fontDPI):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
(webkitWebViewDispose):
(webkit_web_view_set_zoom_level):
(webkit_web_view_get_zoom_level):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(_WebKitWebViewBasePrivate::_WebKitWebViewBasePrivate):
(refreshInternalScaling):
(webkitWebViewBaseUpdateDisplayID):
(webkitWebViewBaseDispose):
(webkitWebViewBaseGetScaleFactors):
(deviceScaleFactorChanged):
(webkitWebViewBaseCreateWebPage):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:
</pre>
https://github.com/WebKit/WebKit/commit/69ad530f10603697a1c53b063840abc8ba7d9efe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49580 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28865 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52627 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52822 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/256 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51884 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34885 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26485 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40474 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26400 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/21589 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/23853 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43888 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7951 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44394 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54399 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24665 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10881 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26778 "Built successfully") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25658 "Failed to checkout and rebase branch from PR 26938") | | |

### CODE REVIEW

**Comment by mcatanzaro** [comment]
Hi, we'll look closer at this next week, but for now please start by splitting this into two pull requests because unfortunately each pull request has to have exactly one commit in order to land. Probably best to squash that second commit into the first one.

I'm going to mark this one as Draft for now. You can create a separate pull request with just one commit (corresponding to the first two commits from this pull request). Then after that pull request lands, you can rebase this one such that it contains only the third commit, and remove the Draft status.

**Comment by gwhitney** [comment]
OK,will do ASAP.

**Comment by gwhitney** [comment]
Filed #26954.

**Comment by mcatanzaro** [comment]
Reminder: after https://github.com/WebKit/WebKit/pull/26954 lands, rebase this pull request and then ping @WebKit/glib-reviewers to re-request review.

**Comment by gwhitney** [comment]
OK, I see that #26954 has landed. I will rebase/update this pull request as soon as I am able and mention the handle you gave when it is done. Thanks.

**Comment by gwhitney** [comment]
This bugfix has been updated with respect to the further discussion in https://bugs.webkit.org/show_bug.cgi?id=247980 and rebased on top of #26954. I believe it resolves the associated issue, and would welcome @bubbleguuum to test it if possible. Thanks in advance for reviewing, @WebKit/glib-reviewers.

File: Source/WebCore/platform/gtk/PlatformScreenGtk.cpp
Comment (mcatanzaro) [inline Source/WebCore/platform/gtk/PlatformScreenGtk.cpp]: Style nit:

```suggestion
    if (auto* display = gdk_display_get_default()) {
        if (auto* screen = gdk_display_get_default_screen(display))
```

File: Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:346]: Any idea what we should do with https://bugs.webkit.org/show_bug.cgi?id=250138, which complains that the text scale factor is used only for text and not also for the rest of the web content?

File: Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp]:
```suggestion
        page->setPageZoomFactor(page->pageZoomFactor() * scaleRatio);
```

File: Source/WebCore/platform/gtk/PlatformScreenGtk.cpp
Comment (mcatanzaro) [inline Source/WebCore/platform/gtk/PlatformScreenGtk.cpp]:
```suggestion
#endif

    static GtkSettings* gtkSettings = gtk_settings_get_default();
```

**Review by mcatanzaro — COMMENTED** [review]
Looks fine to me. Just a few nits regarding style.

File: Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
Comment (gwhitney) [reply Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:346]: Ah, I hadn't been aware of those reports. Give me a couple of days to reproduce them and see how this change affects them. Ultimately it should always be possible to get decent-looking results with this PR, but yes, it does give two "knobs" because GTK is set up that way with two notions of scaling, and the font scaling, if I am understanding GTK documentation and the behavior of other GTK applications correctly, _is_ intended only to scale fonts, leaving layouts and geometry unaffected. (There are lots of other GTK applications besides those built with WebKitGTK that suffer from text being drawn outside the space allocated for it with some combinations of screen and text scaling, believe me!) But I would like to make a coherent response to 250138; I will likely post something there referring to this PR. In the meantime, would you like me to mark this PR as draft?

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:346]: I'll let you decide. I'm fine with landing this now and leaving 250138 open for follow-up. Maybe it's better to wait a few days for you to investigate, though. Thanks for your help with this regardless.

Comment (gwhitney) [reply Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:346]: OK, marking as draft for now then.

**Review by mcatanzaro — APPROVED** [review]
Thanks for your comment in https://bugs.webkit.org/show_bug.cgi?id=250138. I'm going to click the Approve button. Time to update with my suggested changes and mark as ready rather than Draft.

**Comment by gwhitney** [comment]
OK, I will amend the commit as requested and unmark as draft within the next 24 hours. Thanks!

File: Source/WebCore/platform/gtk/PlatformScreenGtk.cpp
Comment (gwhitney) [reply Source/WebCore/platform/gtk/PlatformScreenGtk.cpp]: Changed as requested.

File: Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
Comment (gwhitney) [reply Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp]: Spaces inserted.

File: Source/WebCore/platform/gtk/PlatformScreenGtk.cpp
Comment (gwhitney) [reply Source/WebCore/platform/gtk/PlatformScreenGtk.cpp]: Added blank line as requested.

**Comment by gwhitney** [comment]
Sorry about the delay resulting from the conversation on webkit bugzilla 250138. Hope this is ready to go now, even if that conversation was unsatisfactory.

**Review by mcatanzaro — CHANGES_REQUESTED** [review]
Unfortunately this breaks the Large Text accessibility setting, see https://bugs.webkit.org/show_bug.cgi?id=250138. We'll need to figure this out before landing.

**Review by mcatanzaro — APPROVED** [review]
Whoops, sorry... I failed to test whether it worked _without_ changes, and it doesn't. Seems the large text setting is ~~only respected when running in flatpak~~ not respected if I fail to change it properly. Let's not block on this.

**Pull request merged**
